database-migration/src/main/java/uk/ac/ebi/eva/dbmigration/mongodb/ExtractAnnotationFromVariant.java

```diff
@@ -0,0 +1,312 @@
+
+    static final String XREF_ID_FIELD = "id";
+
+    private static DatabaseParameters databaseParameters;
+
+    public static void setDatabaseParameters(DatabaseParameters databaseParameters) {
+        ExtractAnnotationFromVariant.databaseParameters = databaseParameters;
+    }
+
+    @ChangeSet(order = "001", id = "migrateAnnotation", author = "EVA")
+    public void migrateAnnotation(MongoDatabase mongoDatabase) {
+        final MongoCollection<Document> variantsCollection = mongoDatabase.getCollection(
+                databaseParameters.getDbCollectionsVariantsName());
+        final MongoCollection<Document> annotationCollection = mongoDatabase.getCollection(
+                databaseParameters.getDbCollectionsAnnotationsName());
+        logger.info("1) migrate annotation from collection {}", variantsCollection.getNamespace());
+
+        long counter = 0;
+        long inserted = 0;
+        BulkWriteOptions unorderedBulk = new BulkWriteOptions().ordered(false);
+        try (MongoCursor<Document> cursor = variantsCollection.find().iterator()) {
+            boolean cursorIsConsumed;
+            while (true) {
+                List<InsertOneModel<Document>> annotationsToInsert = getBatch(cursor, BULK_SIZE, this::buildInsertion);
+                cursorIsConsumed = annotationsToInsert.isEmpty();
+                if (cursorIsConsumed) {
+                    break;
+                }
+
+                counter += annotationsToInsert.size();
+                BulkWriteResult bulkInsert = annotationCollection.bulkWrite(annotationsToInsert, unorderedBulk);
+                inserted += bulkInsert.getInsertedCount();
+            }
+        }
+
+        //before executing the next changeSet check that the count of read and written annotation documents match
+        if (counter != inserted) {
+            throw new RuntimeException(
+                    "The number of processed Variants (" + counter + ") is different from the number of new annotation "
+                            + "inserted (" + inserted + "). The '" + ANNOT_FIELD + "' field will not be removed "
+                            + "from the " + variantsCollection.getNamespace() + " collection.");
+        }
+    }
+
+    /**
+     * Return a batch of Documents processed, advancing the cursor provided.
+     * @param cursor won't be closed, please close it outside this function.
+     * @param bulkSize maximum size for the batch. The list returned can be smaller.
+     * @param transformation function to apply to each document in the cursor, and accumulate the result in the list.
+     * @return A list with processed documents, or an empty list if there are no more elements in the cursor.
+     */
+    private <T> List<T> getBatch(MongoCursor<Document> cursor, int bulkSize, Function<Document, T> transformation) {
+        List<T> batch = new ArrayList<>();
+        int counter = 0;
+        while (cursor.hasNext()) {
+            T element = transformation.apply(cursor.next());
```

Comment:
do you guys think this function is doing too many things? maybe I could do the transformation outside, but I think this version is more useful as it has a small and specific scope